plutus-chain-index/app/Marconi/Index/ScriptTx.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TemplateHaskell            #-}
+
+
+module Marconi.Index.ScriptTx where
+
+import Cardano.Api (SlotNo)
+import Codec.Serialise (serialise)
+import Control.Lens.Operators ((^.))
+import Control.Lens.TH (makeLenses)
+import Data.ByteString qualified as BS
+import Data.ByteString.Lazy (toStrict)
+import Data.Foldable (forM_)
+import Data.Maybe (fromJust)
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.FromField qualified as SQL
+import Database.SQLite.Simple.ToField qualified as SQL
+import GHC.Generics (Generic)
+import Index.VSqlite (SqliteIndex)
+import Index.VSqlite qualified as Ix
+import Ledger (ScriptHash)
+
+
+newtype Depth = Depth Int
+
+newtype ScriptAddress = ScriptAddress ScriptHash
+  deriving (Show, Eq)
+newtype TxCbor = TxCbor BS.ByteString
+  deriving (Show)
+  deriving newtype (SQL.ToField, SQL.FromField)
+
+-- * SQLite
+
+data ScriptTxRow = ScriptTxRow
+  { _scriptAddress :: !ScriptAddress
+  , _txCbor        :: !TxCbor
+  } deriving (Generic)
+
+$(makeLenses ''ScriptTxRow)
+
+instance SQL.ToField ScriptAddress where
+  toField (ScriptAddress hash)  = SQL.SQLBlob . toStrict . serialise $ hash
+instance SQL.FromField ScriptAddress where
+  fromField _ = undefined -- todo
+
+instance SQL.ToRow ScriptTxRow where
+  toRow o = [SQL.toField $ o ^. scriptAddress, SQL.toField $ o ^. txCbor]
+
+-- * Indexer
+
+type Query = ScriptAddress
+type Result = [TxCbor]
+
+data ScriptTxUpdate = ScriptTxUpdate
+  { _txScripts :: [(TxCbor, [ScriptAddress])]
+  , _slotNo    :: !SlotNo
+  } deriving (Show)
+
+$(makeLenses ''ScriptTxUpdate)
+
+type ScriptTxIndex = SqliteIndex ScriptTxUpdate () Query Result
+
+open :: FilePath -> Depth -> IO ScriptTxIndex
+open dbPath (Depth k) = do
+  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 2) dbPath
+  let c = ix ^. Ix.handle
+  SQL.execute_ c "CREATE TABLE IF NOT EXISTS script_transactions (scriptAddress TEXT NOT NULL, txCbor BLOB NOT NULL)"
```

Comment:
I think I would rather have a primary key index on `scriptAddress`. Is there a reason why we are not doing that?

Reply:
I might be wrong here, but since a script can be targeted by many transactions (?), then the `scriptAddress` won't be unique and so it can't be a primary key. (Though it still can have a index on it though, so it would be faster to search for with sql).

Reply:
(The `scriptAddress` primary key issue is also mentioned in the PR description.)